docs/architecture-decisions/2024-05-22_teal-compilation.md

```diff
@@ -0,0 +1,84 @@
+# Compilation of Algorand TEAL Programs into AVM Bytecode
+
+- **Status**: Proposed
+- **Owner**: Daniel McGregor (MakerX)
+- **Deciders**: Rob Moore (MakerX), Adam Chidlow (MakerX), Alessandro Cappellato (Algorand Foundation)
+- **Date created**: 2024-05-22
+- **Date decided**: 
+- **Date updated**: 
+
+## Context
+
+The Puya compiler needs to assemble valid Algorand TEAL (Transaction Execution Approval Language) programs into AVM (Algorand Virtual Machine) bytecode 
+so the bytecode can be used within inner transactions that need to create applications.
```

Comment:
Minor: That's the driver of it, but there are naturally other areas where having that capability is useful, particularly with ARC-56 coming where if there aren't template variables we can precompile for you and avoid an extra algod call in consuming applications.